tiledb/api/src/query/conditions.rs

```diff
@@ -0,0 +1,736 @@
+        })
+    }
+
+    pub fn negate(&self) -> TileDBResult<Self> {
+        let mut c_cond: *mut ffi::tiledb_query_condition_t = out_ptr!();
+        self.capi_return(unsafe {
+            ffi::tiledb_query_condition_negate(
+                self.context.capi(),
+                self.capi(),
+                &mut c_cond,
+            )
+        })?;
+
+        Ok(QueryCondition {
+            context: self.context,
+            raw: RawQueryCondition::Owned(c_cond),
+        })
+    }
+}
+
+pub trait QueryConditionInitializer {
+    fn initialize<'ctx>(
+        &self,
+        ctx: &'ctx Context,
+    ) -> TileDBResult<QueryCondition<'ctx>>;
+}
+
+pub struct ComparisonQueryConditionInitializer<V: CAPIConverter> {
+    field: String,
+    value: V,
+    op: ComparisonOperation,
+}
+
+impl<V: CAPIConverter + Clone> QueryConditionInitializer
+    for ComparisonQueryConditionInitializer<V>
+{
+    fn initialize<'ctx>(
+        &self,
+        ctx: &'ctx Context,
+    ) -> TileDBResult<QueryCondition<'ctx>> {
+        QueryCondition::new(ctx, &self.field, self.value.clone(), self.op)
+    }
+}
+
+pub struct StringComparisonQueryConditionInitializer {
+    field: String,
+    value: String,
+    op: ComparisonOperation,
+}
+
+impl QueryConditionInitializer for StringComparisonQueryConditionInitializer {
+    fn initialize<'ctx>(
+        &self,
+        ctx: &'ctx Context,
+    ) -> TileDBResult<QueryCondition<'ctx>> {
+        QueryCondition::new_string(ctx, &self.field, &self.value, self.op)
+    }
+}
+
+pub struct SetMembershipQueryConditionInitializer {
+    field: String,
+    data: Vec<u8>,
+    offsets: Vec<u64>,
+    op: SetMembershipOperation,
+}
+
+impl QueryConditionInitializer for SetMembershipQueryConditionInitializer {
+    fn initialize<'ctx>(
+        &self,
+        ctx: &'ctx Context,
+    ) -> TileDBResult<QueryCondition<'ctx>> {
+        QueryCondition::new_set(
+            ctx,
+            &self.field,
+            &self.data,
+            &self.offsets,
+            self.op,
+        )
+    }
+}
+
+pub struct CombinationQueryConditionInitializer {
+    lhs: Box<dyn QueryConditionInitializer>,
+    rhs: Box<dyn QueryConditionInitializer>,
+    op: CombinationOperation,
+}
+
+impl QueryConditionInitializer for CombinationQueryConditionInitializer {
+    fn initialize<'ctx>(
+        &self,
+        ctx: &'ctx Context,
+    ) -> TileDBResult<QueryCondition<'ctx>> {
+        let lhs = self.lhs.initialize(ctx)?;
+        let rhs = self.rhs.initialize(ctx)?;
+        lhs.combine(&rhs, self.op)
+    }
+}
+
+pub struct NegationQueryConditionInitializer {
+    initializer: Box<dyn QueryConditionInitializer>,
+}
+
+impl QueryConditionInitializer for NegationQueryConditionInitializer {
+    fn initialize<'ctx>(
+        &self,
+        ctx: &'ctx Context,
+    ) -> TileDBResult<QueryCondition<'ctx>> {
+        let cond = self.initializer.initialize(ctx)?;
+        cond.negate()
+    }
+}
+
+pub struct QC {
```

Comment:
This is a variable right?  It should be named variable

Reply:
I'm not sure what you mean by this. I assume the `new` method was a bit misleading though. A QC is meant to represent any query condition whether it be a simple `x < 5` comparison or some complicated nested tree of boolean algebra. The new implementation has a hopefully less confusing API that looks similar but with better types.

At the bottom of the type tree is a Condition which has the basic shape of `field operator value`. There are three variants to enforce correctness:

1. CmpCondition - field {<, <=, ==, !=, >=, >} value which is a single value
2. SetMemberCondition field {in, not_in} value which is a vec of values
3. NullCondition: field {is_null, not_null} (value elided cause its nullptr)

After conditions, we have the `QC` structure which has three variants:

1. Cond - One of the three condition types from above
2. Comb (lhs: QC, rhs: QC, operator) <- two sub conditions with a combination operator (AND or OR)
3. Negate (val: QC) <- Applies !QC

The final bit is the syntactical sugar for creating and combining query conditions. Creating a condition starts with `QC::field("foo")` which returns a `Field` struct. A `Field` just contains a copy of the field name in a `String` and provides a number of helper methods to generate any of the three variant types. Thus you can do any of the following:

```rust
let qc1 = QC::field("foo").lt(5);
let qc2 = QC::field("bar").in(&[1, 2, 3, 4][..]);
let qc3 = QC::field("baz").is_null();
```

Then for combining and negating, there are the `BitAnd`, `BitOr`, and `Not` which allow for things like:

```rust
let qc4 = qc1 & qc2;
let qc5 = qc2 | qc3;
let qc6 = !qc5;
```

When we get to predicate push down, we'll also probably add more programatic methods for conversion. Something like the following. I've not written these yet though because I don't want to add them before we need them as well as I could see slight variations depending on what the other side of the conversion process looks like.

```rust
let qc1 = QC::field("foo").cmp(CmpOp::Less, 5);
let qc2 = QC::field("bar").set_member(SetMemberOp::In, &[1, 2, 3, 4, 5][..]);
let qc3 = QC::field("baz").nullness(NullnessOp::IsNull);
let qc4 = qc1.combine(qc2, CombOp::And);
let qc5 = qc2.combine(qc3, CombOp::Or);
let qc6 = qc5.negate();
```